packages/gatsby/src/utils/page-data.ts

```diff
@@ -76,15 +78,24 @@ export async function writePageData(
     `json`,
     `${pagePath.replace(/\//g, `_`)}.json`
   )
+
   const outputFilePath = getFilePath(publicDir, pagePath)
   const result = await fs.readJSON(inputFilePath)
   const body = {
     componentChunkName,
     path: pagePath,
     matchPath,
+    notInDevBundle,
     result,
     staticQueryHashes,
   }
+  if (process.env.gatsby_executing_command === `develop`) {
+    const { clientVisitedPages } = store.getState()
+    body.notInDevBundle = !clientVisitedPages.has(componentChunkName)
+  } else {
+    delete body.notInDevBundle
+  }
+
```

Comment:
This doesn't guarantee that `notInDevBundle` will not end up in production artefacts, as we do reuse `page-data.json` written in develop - if you run `gatsby develop`, stop it after it's done and run `gatsby build` afterwards you will likely see this field there

Reply:
yeah — I realize this. I consider this acceptable as:

1. it wouldn't affect the production site and it's a tiny amount of extra data if it does end up in production page-data.json files
2. most people have a separate build server so wouldn't run into this scenario

If there's a different way we can think of to pass the state of whether a path's page component is in the bundle or not that'd be great. E.g. I thought of exposing a route for the runtime to query for this. But honestly this feels like perhaps the cleanest. It ties together writing out the staticQueryHashes w/ changing this boolean. Any other solution would have race conditions.